api/types/databaseobject/databaseobject_test.go

```diff
@@ -0,0 +1,148 @@
+					Name:      "valid object",
+					Namespace: defaults.Namespace,
+				},
+				Spec: &dbobjectv1.DatabaseObjectSpec{
+					Protocol:      "postgres",
+					DbServiceName: "test",
+					ObjectKind:    types.KindDatabaseObject,
+					Database:      "test",
+					Schema:        "test",
+					Name:          "test",
+				},
+			},
+		},
+		{
+			name:          "missing name",
+			spec:          &dbobjectv1.DatabaseObjectSpec{},
+			expectedError: trace.BadParameter("spec.name: must be non-empty"),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			obj, err := NewDatabaseObject(tt.name, tt.spec)
+			require.ErrorIs(t, err, tt.expectedError)
+			require.Equal(t, tt.expectedObj, obj)
+		})
+	}
+}
+
+func TestValidateDatabaseObject(t *testing.T) {
+	tests := []struct {
+		name           string
+		databaseObject *dbobjectv1.DatabaseObject
+		expectedError  error
+	}{
+		{
+			name: "valid object",
+			databaseObject: &dbobjectv1.DatabaseObject{
+				Kind:     types.KindDatabaseObject,
+				Metadata: &headerv1.Metadata{Name: "test", Namespace: defaults.Namespace},
+				Spec:     &dbobjectv1.DatabaseObjectSpec{Name: "test", Protocol: "test"},
+			},
+			expectedError: nil,
+		},
+		{
+			name:           "nil object",
+			databaseObject: nil,
+			expectedError:  trace.BadParameter("object must be non-nil"),
+		},
+		{
+			name:           "nil metadata",
+			databaseObject: &dbobjectv1.DatabaseObject{Metadata: nil},
+			expectedError:  trace.BadParameter("metadata: must be non-nil"),
+		},
+		{
+			name: "invalid kind",
+			databaseObject: &dbobjectv1.DatabaseObject{
+				Kind:     "InvalidKind",
+				Metadata: &headerv1.Metadata{Name: "test", Namespace: defaults.Namespace},
+				Spec:     &dbobjectv1.DatabaseObjectSpec{Name: "test"},
+			},
+			expectedError: trace.BadParameter("wrong kind %v", "InvalidKind"),
+		},
+		{
+			name: "missing spec",
+			databaseObject: &dbobjectv1.DatabaseObject{
+				Kind:     types.KindDatabaseObject,
+				Metadata: &headerv1.Metadata{Name: "test", Namespace: defaults.Namespace},
+				Spec:     nil,
+			},
+			expectedError: trace.BadParameter("spec: must be non-empty"),
+		},
+		{
+			name: "missing object name",
+			databaseObject: &dbobjectv1.DatabaseObject{
+				Kind:     types.KindDatabaseObject,
+				Metadata: &headerv1.Metadata{Name: "", Namespace: defaults.Namespace},
+			},
+			expectedError: trace.BadParameter("metadata.name: must be non-empty"),
+		},
+		{
+			name: "missing name",
+			databaseObject: &dbobjectv1.DatabaseObject{
+				Kind:     types.KindDatabaseObject,
+				Metadata: &headerv1.Metadata{Name: "test", Namespace: defaults.Namespace},
+				Spec:     &dbobjectv1.DatabaseObjectSpec{Name: ""},
+			},
+			expectedError: trace.BadParameter("spec.name: must be non-empty"),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := ValidateDatabaseObject(tt.databaseObject)
+			require.ErrorIs(t, err, tt.expectedError)
```

Comment:
Ditto for ErrorsIs. I suggest a combination of ErrorContains and either ErrorAs(new(*trace.BadParameterError)) or a trace.IsBadParameter check.

Reply:
I disagree this is a problem, see my argument in the other comment.